### PR TITLE
Initial change tracking for skip navigations

### PR DESCRIPTION
**Pull request by ajcvickers**
Part of #19003

Covers only the simple case of collection skips with non-composite, non-alternate keys and simple concrete join table containing only two FK properties.

Many-to-many tracking queries are enabled, but some are skipped due to query translation errors.

### CODE REVIEW

File: src/EFCore/Diagnostics/NavigationBaseEventData.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+            [NotNull] Func<EventDefinitionBase, EventData, string> messageGenerator,
+            [NotNull] INavigationBase navigationBase)
+            : base(eventDefinition, messageGenerator)
+        {
+        }
+
+        /// <summary>
+        ///     The navigation.
+        /// </summary>
+        public abstract INavigationBase NavigationBase { get; }
```
Comment (AndriySvyryd) [inline src/EFCore/Diagnostics/NavigationBaseEventData.cs:33]: Is this property necessary? And this type in general?

File: src/EFCore/Metadata/IConstructorBindingFactory.cs [hunk 2]
```diff
@@ -30,7 +30,6 @@ public interface IConstructorBindingFactory
         /// <param name="binding"> The binding, or <see langword="null" /> if <see langword="null" /> could be created. </param>
         /// <param name="unboundParameters"> The parameters that could not be bound. </param>
         /// <returns> <see langword="true" /> if a binding was created; <see langword="false" /> otherwise. </returns>
-        [ContractAnnotation("=>true, binding:notnull, failedBindings:null; =>false, binding:null, failedBindings:notnull")]
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/IConstructorBindingFactory.cs:33]: Why are these removed?

File: src/EFCore/ChangeTracking/Internal/NavigationFixer.cs [hunk 3]
```diff
@@ -732,32 +897,119 @@ private void DeleteFixup(InternalEntityEntry entry)
...
+            }
+        }
+
+        private static InternalEntityEntry FindOrCreateAssociationEntry(
+            InternalEntityEntry entry, InternalEntityEntry otherEntry, ISkipNavigation skipNavigation,
+            bool fromQuery, bool setModified)
+        {
+            var associationEntry = FindAssociationEntry(entry, otherEntry, skipNavigation);
+
+            if (associationEntry == null)
```
Comment (AndriySvyryd) [inline src/EFCore/ChangeTracking/Internal/NavigationFixer.cs:952]: I though we agreed to limit this to entity types that only have the foreign keys

File: src/EFCore/Diagnostics/NavigationBaseEventData.cs (hunk 1 above)
Comment (ajcvickers) [reply src/EFCore/Diagnostics/NavigationBaseEventData.cs:33]: I didn't want to break the existing event data because not breaking across versions was discussed a lot in the original design. However, I wanted people to be able to write common code for processing any navigation event. But I don't feel strongly about it.

File: src/EFCore/Metadata/IConstructorBindingFactory.cs (hunk 2 above)
Comment (ajcvickers) [reply src/EFCore/Metadata/IConstructorBindingFactory.cs:33]: Because they are wrong. At some point the code was changed, but the contracts were not updated. Are we doing these kind of annotations in the code in general? If so, we should figure out how to ensure they stay up-to-date.

File: src/EFCore/ChangeTracking/Internal/NavigationFixer.cs (hunk 3 above)
Comment (ajcvickers) [reply src/EFCore/ChangeTracking/Internal/NavigationFixer.cs:952]: I'm not sure I 100% agreed. ;-) Anyway, I'm going to look into this more, so the code here is not final.

File: src/EFCore/ChangeTracking/Internal/NavigationFixer.cs [hunk 4]
```diff
@@ -234,20 +242,39 @@ public virtual void NavigationReferenceChanged(InternalEntityEntry entry, INavig
...
+                        {
+                            var associationEntry = FindAssociationEntry(entry, oldTargetEntry, skipNavigation);
+                            associationEntry?.SetEntityState(EntityState.Deleted);
 
-                        if (inverse != null
-                            && ReferenceEquals(oldTargetEntry[inverse], entry.Entity)
-                            && (!foreignKey.IsOwnership
-                                || (oldTargetEntry.EntityState != EntityState.Deleted
-                                    && oldTargetEntry.EntityState != EntityState.Detached)))
+                            if (skipNavigation.Inverse.IsCollection)
```
Comment (lajones) [inline src/EFCore/ChangeTracking/Internal/NavigationFixer.cs:250]:
```suggestion
                            if (skipNavigation.Inverse?.IsCollection)
```

Although it's weird, I had some tests at one point that managed to set up skipNav's with strange inverses including `null`. So suggest the `?` form. Similarly in a few places below. That said, in this particular case you may be able to ignore the inverse and carry on rather than throwing in the `else`?

Comment (ajcvickers) [reply src/EFCore/ChangeTracking/Internal/NavigationFixer.cs:250]: See discussion from last week in team chat. @AndriySvyryd asserted that the Inverse will not be null. It may be that the cases you were hitting were while the model was being built.

Comment (AndriySvyryd) [reply src/EFCore/ChangeTracking/Internal/NavigationFixer.cs:250]: Inverse will always be non-null after model validation

File: src/EFCore/ChangeTracking/Internal/InternalEntityEntry.cs [hunk 5]
```diff
@@ -825,11 +831,13 @@ public virtual bool CollectionContains([NotNull] INavigation navigation, [NotNul
         ///     any release. You should only use it directly in your code with extreme caution and knowing that
         ///     doing so can result in application failures when updating to a new Entity Framework Core release.
         /// </summary>
-        public virtual bool RemoveFromCollection([NotNull] INavigation navigation, [NotNull] InternalEntityEntry value)
+        public virtual bool RemoveFromCollection([NotNull] INavigationBase navigation, [NotNull] InternalEntityEntry value)
         {
             Check.DebugAssert(!navigation.IsShadowProperty(), "navigation is shadow property");
 
-            return ((Navigation)navigation).CollectionAccessor.Remove(Entity, value.Entity);
+            return navigation is SkipNavigation skipNavigation
```
Comment (smitpatel) [inline src/EFCore/ChangeTracking/Internal/InternalEntityEntry.cs:838]: Consider DRYing
- Either private static method in this class to get CollectionAccessor for INavigationBase
- or if used at other places too then extension method.

File: src/EFCore/Diagnostics/NavigationEventData.cs [hunk 6]
```diff
@@ -9,10 +9,9 @@
 namespace Microsoft.EntityFrameworkCore.Diagnostics
 {
     /// <summary>
-    ///     A <see cref="DiagnosticSource" /> event payload class for events that have
-    ///     a navigation.
+    ///     A <see cref="DiagnosticSource" /> event payload class for events that have an <see cref="ISkipNavigation"/>.
```
Comment (smitpatel) [inline src/EFCore/Diagnostics/NavigationEventData.cs:12]: Is this right?

File: src/EFCore/Diagnostics/SkipNavigationEventData.cs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+using System;
+using System.Diagnostics;
+using JetBrains.Annotations;
+using Microsoft.EntityFrameworkCore.Metadata;
+
+namespace Microsoft.EntityFrameworkCore.Diagnostics
+{
+    /// <summary>
+    ///     A <see cref="DiagnosticSource" /> event payload class for events that have an <see cref="INavigation"/>.
```
Comment (smitpatel) [inline src/EFCore/Diagnostics/SkipNavigationEventData.cs:12]: 👀

File: test/EFCore.Specification.Tests/ManyToManyTrackingTestBase.cs [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+        public virtual void Can_delete_with_many_to_many()
+        {
+            ExecuteWithStrategyInTransaction(
+                context =>
+                {
+                    var ones = context.Set<EntityOne>().Include(e => e.TwoSkip).OrderBy(e => e.Id).ToList();
+                    var twos = context.Set<EntityTwo>().Include(e => e.OneSkip).ToList();
+
+                    // Make sure other related entities are loaded for delete fixup
+                    context.Set<EntityThree>().Load();
```
Comment (smitpatel) [inline test/EFCore.Specification.Tests/ManyToManyTrackingTestBase.cs:240]: Is this just to verify delete fixup code or needed by customers also?

File: src/EFCore/Diagnostics/NavigationBaseEventData.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore/Diagnostics/NavigationBaseEventData.cs:33]: Using an explicitly-implemented interface would be better if you want to enable common processing

File: src/EFCore/Metadata/IConstructorBindingFactory.cs (hunk 2 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/IConstructorBindingFactory.cs:33]: Well, you can make R# warnings to be errors, but it'll take a while to fix them. I'm fine with removing the ones that are wrong

File: test/EFCore.Specification.Tests/ManyToManyTrackingTestBase.cs (hunk 8 above)
Comment (ajcvickers) [reply test/EFCore.Specification.Tests/ManyToManyTrackingTestBase.cs:240]: It's because there are dependents in the database and if they are not loaded then EF will not null the FK values and hence there will be a database constraint violation when saving. This just avoids that happening in this test--it's purely a test artifact.

**Comment by ajcvickers** [comment]
@smitpatel Any more comments on this?

File: test/EFCore.Sqlite.FunctionalTests/Query/ManyToManyQuerySqliteTest.cs [hunk 9]
```diff
@@ -15,11 +15,125 @@ public ManyToManyQuerySqliteTest(ManyToManyQuerySqliteFixture fixture)
...
+
+        [ConditionalTheory(Skip = "TODO: query translation #19003")]
+        public override Task Filtered_include_skip_navigation_where_split(bool async)
+            => Task.CompletedTask;
+
+        [ConditionalTheory(Skip = "TODO: query translation #19003")]
+        public override Task Filtered_include_skip_navigation_where_then_include_skip_navigation_order_by_skip_take_split(bool async)
+            => Task.CompletedTask;
+
+        [ConditionalTheory(Skip = "TODO: query translation #19003")]
```
Comment (smitpatel) [inline test/EFCore.Sqlite.FunctionalTests/Query/ManyToManyQuerySqliteTest.cs:72]: These are still not working?

Comment (ajcvickers) [reply test/EFCore.Sqlite.FunctionalTests/Query/ManyToManyQuerySqliteTest.cs:72]: Didn't realize you had merged the fix. Will check.

Comment (ajcvickers) [reply test/EFCore.Sqlite.FunctionalTests/Query/ManyToManyQuerySqliteTest.cs]: All passing. 👍

File: src/EFCore/ChangeTracking/Internal/NavigationFixer.cs (hunk 3 above)
Comment (AndriySvyryd) [reply src/EFCore/ChangeTracking/Internal/NavigationFixer.cs:952]: The description says:
> Covers only the simple case of collection skips with non-composite, non-alternate keys and simple concrete join table containing only two FK properties

Comment (ajcvickers) [reply src/EFCore/ChangeTracking/Internal/NavigationFixer.cs:952]: I talked to your manager, and he said it's okay to merge this now and revisit later.

Comment (AndriySvyryd) [reply src/EFCore/ChangeTracking/Internal/NavigationFixer.cs:952]: Perhaps I should speak to your manager then 🙃
